README.md

[![Build Status](https://img.shields.io/travis/sclark007/chef_repo_server.svg)](https://travis-ci.org/sclark007/chef_repo_server)
[![Code Climate](https://codeclimate.com/github/sclark007/chef_repo_server/badges/gpa.svg)](https://codeclimate.com/github/sclark007/chef_repo_server)
[![Coverage Status](https://img.shields.io/coveralls/sclark007/chef_repo_server.svg)](https://coveralls.io/github/sclark007/chef_repo_server)


chef_repo_server
================

To often when we build chef automation pipelines and we need access to a gem api and the chef rpm repositories. this cookbook is intended to setup both.

Requirements
============

## Platforms

## chef
- Chef 12.1+

## cookbooks
- `cron` add reposync and create repo tasks to cron
- `ssl_certificate` configure apache with ssl support
- `apache2` https service.

Usage
=====

Using this cookbook is relatively straightforward. It is recommended to create
a project or organization specific [wrapper cookbook](https://www.chef.io/blog/2013/12/03/doing-wrapper-cookbooks-right/)
and add the desired recipes to the run list of a node, or create a role. Depending on your
environment, you may have multiple roles that use different recipes
from this cookbook. Adjust any attributes as desired. For example, to
create a basic role for web servers that provide both HTTP and HTTPS:

```ruby
    % cat recipes/server.rb
    chef_repo_server_mirror 'chef-7' do
      repo_name 'chef-7'
      repo_description 'Chef Packages for Enterprise Linux 7'
      repo_baseurl "#{node['chef_repo_server']['repo_server']}/repos/yum/stable/el/7/x86_64/"
      action :create
    end
```

For examples of using the definitions in your own recipes, see their
respective sections below.

Attributes
==========

- `['chef_repo_server']['gems']` list of gems to be added to the config.ru
- `['chef_repo_server']['basepath']` base directory for storing gems and chef rpms.
- `['chef_repo_server']['repo_server']` server were chef rpm's can be downloaded

# Resources
## chef_repo_server_mirror
- `local_path` location
- `repo_name` repo name
- `repo_description` descrption
- `repo_baseurl` repo url


**Author: ** Steve Clark <steve@bigsteve.us>  
** Copyright (c) ** 2016 Steve Clark  

```
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md


## 0.1.0
create a gem-server for internal systems to use.


test/recipes/default_test.rb
# encoding: utf-8

# Inspec test for recipe chef_repo_server::default

# The Inspec reference, with examples and extensive documentation, can be
# found at https://docs.chef.io/inspec_reference.html

describe port(443) do
  it { should be_listening }
end
describe port(9443) do
  it { should be_listening }
end


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: tsa_gemirro
# Spec:: default
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

# require 'spec_helper'
#
# describe 'tsa_gem_server::default' do
#   before do
#     Fauxhai.mock(platform: 'centos', version: '7.2.1511')
#   end
#
#   let(:chef_run) do
#     runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.2.1511')
#     runner.converge(described_recipe)
#   end
#
#   it 'converges successfully' do
#     expect { chef_run }.to_not raise_error
#   end
#
#   it 'installs httpd' do
#     expect(chef_run).to install_package 'httpd-2.4.6'
#   end
#
#   it 'enables the httpd service' do
#     expect(chef_run).to enable_service 'httpd'
#   end
#
#   it 'starts the httpd service' do
#     expect(chef_run).to start_service 'httpd'
#   end
# end


recipes/_https.rb
# encoding: utf-8
# Cookbook Name:: chef_repo_server
# Recipe:: default
#
#
# Author:: Steve Clark <steve@bigsteve.us>
# Copyright (c) 2016 Steve Clark
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# we need to save the resource variable to get the key and certificate file
# paths
cert = ssl_certificate 'gem-server' do
  common_name node['chef_repo_server']['common_name']
  # common_name '127.0.0.1'
  # we want to be able to use node['my-webapp'] to configure the certificate
  namespace node['my-webapp']
  notifies :restart, 'service[apache2]'
end

include_recipe 'apache2'
include_recipe 'apache2::mod_ssl'

web_app 'chef-server' do
  port '443'
  # this cookbook includes a virtualhost template for apache2
  cookbook 'ssl_certificate'
  server_name cert.common_name
  docroot "#{node['chef_repo_server']['basepath']}/chef"
  directory_options 'Indexes FollowSymLinks MultiViews'
  ssl_key cert.key_path
  ssl_cert cert.cert_path
  ssl_chain cert.chain_path
end

web_app 'gem-server' do
  port '9443'
  # this cookbook includes a virtualhost template for apache2
  cookbook 'ssl_certificate'
  server_name cert.common_name
  docroot "#{node['chef_repo_server']['basepath']}/gems/public/"
  directory_options 'Indexes FollowSymLinks MultiViews'
  ssl_key cert.key_path
  ssl_cert cert.cert_path
  ssl_chain cert.chain_path
end


resources/mirror.rb

# encoding: utf-8
# Cookbook Name:: chef_repo_server
# Recipe:: default
#
#
# Author:: Steve Clark <steve@bigsteve.us>
# Copyright (c) 2016 Steve Clark
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

property :name, [String, Symbol], required: true, name_property: true
property :local_path, String, required: false
property :repo_name, String, required: true
property :repo_description, String, required: true
property :repo_baseurl, String, required: true

default_action :create

def real_local_path
  if local_path == NilClass
    "#{local_path}/#{name}/"
  else
    "#{node['chef_repo_server']['basepath']}/"
  end
end

# Needs to be def'd as path for zap to work.
def path
  "/etc/reposync.repos.d/#{repo_name}.repo"
end
action :create do
  directory '/etc/reposync.repos.d/' do
    owner 'root'
    group 'root'
    mode '0755'
    action :create
  end

  cookbook_file '/etc/reposync.conf' do
    cookbook 'chef_repo_server'
    source 'reposync.conf'
    owner 'root'
    group 'root'
    mode '0644'
    action :create
  end

  template path do
    cookbook 'chef_repo_server'
    source 'repo.erb'
    owner 'root'
    group 'root'
    mode '0644'
    variables(
      repo_name: repo_name,
      repo_description: repo_description,
      repo_baseurl: repo_baseurl
    )
    notifies :run, "execute[#{repo_name}-reposync-once]", :delayed
    action :create
  end
  directory real_local_path do
    owner 'root'
    group 'root'
    mode '0755'
    recursive true
    action :create
  end

  cron_d "#{repo_name}-reposync" do
    predefined_value '@hourly'
    command "reposync -n -d -c /etc/reposync.conf -r #{repo_name} -p #{real_local_path}"
    user    'root'
  end

  cron_d "#{repo_name}-createrepo" do
    predefined_value '@hourly'
    command "createrepo -d -s sha -C #{real_local_path}#{repo_name}"
    user    'root'
  end

  execute "#{repo_name}-reposync-once" do
    command "reposync -n -d -c /etc/reposync.conf -r #{repo_name} -p #{real_local_path}"
    action :nothing
  end
end

# rubocop:enable Metrics/BlockLength
action :delete do
  file path do
    action :delete
  end
  directory real_local_path do
    recursive true
    action :delete
  end
  yum_repository repo_name do
    action :delete
  end
end


recipes/_chef_repo.rb
# encoding: utf-8
# Cookbook Name:: chef_repo_server
# Recipe:: default
#
#
# Author:: Steve Clark <steve@bigsteve.us>
# Copyright (c) 2016 Steve Clark
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

node['chef_repo_server']['rpms'].each do |y|
  yum_package y do
    action :install
  end
end

directory "#{node['chef_repo_server']['basepath']}/chef" do
  owner 'root'
  group 'root'
  mode 0o0755
  recursive true
  action :create
end


recipes/_gem_mirror.rb
# encoding: utf-8
# Cookbook Name:: chef_repo_server
# Recipe:: default
#
#
# Author:: Steve Clark <steve@bigsteve.us>
# Copyright (c) 2016 Steve Clark
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

include_recipe 'cron'

# install the gem
chef_gem 'gem-mirror' do
  action :install
  compile_time true
end

directory "#{node['chef_repo_server']['basepath']}/gems/public/gems" do
  owner 'root'
  group 'root'
  mode 0o0755
  recursive true
  action :create
end

# set config file
template "#{node['chef_repo_server']['basepath']}/gems/config.rb" do
  source 'config.rb.erb'
  owner 'root'
  group 'root'
  variables(
    packages: node['chef_repo_server']['gems']
  )
  mode 0o0744
  notifies :run, 'execute[gem-mirror_update]', :delayed
  notifies :run, 'execute[gem-mirror_index]', :delayed
end

cron_d 'update-gem-mirror' do
  minute  0
  hour    23
  predefined_value '@hourly'
  home '/repos/gems'
  command '/opt/chef/embedded/bin/gem-mirror update && /opt/chef/embedded/bin/gem-mirror index'
  user    'root'
end

execute 'gem-mirror_update' do
  cwd '/repos/gems'
  command '/opt/chef/embedded/bin/gem-mirror update'
  action :nothing
end

execute 'gem-mirror_index' do
  cwd '/repos/gems'
  command '/opt/chef/embedded/bin/gem-mirror index'
  action :nothing
end


attributes/default.rb
# encoding: utf-8
#
# Author:: Steve Clark <steve@bigsteve.us>
# Copyright (c) 2016 Steve Clark
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# rpms needed for creating a yum repo
default['chef_repo_server']['rpms'] = [
  'createrepo',
  'yum-utils',
]

# install lastest version of gems
default['chef_repo_server']['gems'] = {
  'chef-ruby-lvm-attrib' => '~> 0.0',
  'chef-ruby-lvm' => '~> 0.0',
  'chef-vault' => '~> 2.9',
  'inspec' => '~> 1.8',
  'mixlib-install' => '~> 2.1',
  'rake' => '~> 12.0',
  'stove' => '~> 4.1',
  'chef-dk' => '~> 1.1',
}

default['chef_repo_server']['basepath'] = '/repos'

# ssl cert setup
default['chef_repo_server']['repo_server'] = 'https://packages.chef.io'
default['chef_repo_server']['common_name'] = node['fqdn']
default['chef_repo_server']['ssl_cert']['source'] = 'self-signed'
default['chef_repo_server']['ssl_key']['source'] = 'self-signed'

## apache
default['apache']['listen'] = ['*:443', '*:9443']
